lib/modules/manager/gradle/parser.ts

```diff
@@ -891,10 +985,37 @@ function tryMatch({
   return null;
 }
 
+function parseInlineScriptFile(
+  scriptFile: string,
+  variables: PackageVariables,
+  recursionDepth: number,
+  packageFile = ''
+): SyntaxHandlerOutput {
+  if (recursionDepth > 0) {
+    logger.warn({ scriptFile }, `Max recursion depth reached`);
+    return null;
+  }
+
+  const scriptFilePath = getSiblingFileName(packageFile, scriptFile);
+  const scriptFileContent = readLocalFileSync(scriptFilePath, 'utf-8');
```

Comment:
i don't like this.

@zharinov any suggestions to get this async

Reply:
Hm, and I also wonder if this creates an attack vector, i.e. for arbitrary file in the machine running Renovate to be read and leaked via log/stacktrace.

Reply:
I mean not necessary, just need to be sure we can't read files outside of `localDir`

Reply:
Really like @Churro contributions, though

Reply:
So, here is my take:
- We include only `*.gradle` files, but users can configure other patterns via custom `fileMatch`
- Process files in the usual order, additionally marking them as visited/unvisited
- When the `apply from` construction is encountered, we perform out-of-order processing (probably, via callback mechanism) and continue as usual
- If we already have visited file and then need to *apply it from*, I guess we need re-visit it and replace all previous results
- If some file has been visited before we process it according our usual order, we skip it

Reply:
Thank you for your comments and for thinking this through!

Regarding the potential attack vector, I've now added safeguards to mitigate this from two sides:
- Only files matching `\.gradle(\.kts)?$` can be used with `apply from`
- The final file path (= after `upath.join()`, which internally normalizes `../../..`) is ensured to be within `localDir`

Regarding @zharinov 's notes:
* Allowing a pattern override via `fileMatch` is not necessary, as there is simply no use-case for it. Although it technically works to include non-gradle files, in practice simply nobody does (see [[1](https://cs.opensource.google/search?q=%22apply%20from:%22)], [[2](https://github.com/search?l=Gradle&p=1&q="apply+from"&ref=advsearch&type=Code)], [[3](https://github.com/search?l=Kotlin&p=1&q="apply+from"&ref=advsearch&type=Code)]). Probably I shouldn't have even mentioned this because it may have raised the idea this would be common, although it's not.
  * AFAIK, `fileMatch` cannot be accessed in the manager and it's not included in `GlobalConfig.get()`, the only singleton I could spot where it may eventually be located. Still, I don't see the need to over-engineer this.
* The outlined approach would be significantly more complex with no memory or runtime benefits whatsoever:
  * I'd need to store all parsing results individually in a Map-like structure instead of merging them ad-hoc. I would this always, not only when `apply from` is used in any package file
  * I'd somehow need to bind to this structure to access it from the callback, recurringly re-visit files (think of multiple `apply from` statements in one gradle file) although marked visited, etc.
  * The existing `await readLocalFile()` already doesn't enjoy the advantages of async functions due to _await_. The need to introduce `readLocalFileSync` stems from the fact that `extractAllPackageFiles()` is marked async, although what it does is happening async in fact
  * The only slight runtime optimization would be if a file is mentioned in two different `apply from` statements. The second time, it would then not need to be parsed again. Still, this is in the range of micro-seconds.

As a sign of good will, I have now tried to implement this for the last ~2h and failed at various points. Going on with this would cost me more hours of engineering and testing efforts to basically end up with the very same result in a way more complex, less intuitively maintainable fashion.

I'd request you to please review the PR in the state it is right now. For sure, I'm ready to address probably other issues you may find.

Reply:
FYI I did this quick test of `upath` to try to assess if it at least resolves the simplest path traversal attack and it does:

<img width="426" alt="image" src="https://user-images.githubusercontent.com/6311784/173736267-afc5bb41-45ec-4418-b17f-d32e54798230.png">

In other words it does not preserve the `../../`

Reply:
Thanks for checking this too. What makes me confident the safeguard works effectively:
- [Documentation on path.join()](https://nodejs.org/api/path.html#pathjoinpaths), which `upath` calls internally confirms this normalization happens on purpos
- It is inline with [official recommendations from nodejs](https://nodejs.org/en/knowledge/file-system/security/introduction/#preventing-directory-traversal) to prevent path traversal attacks

Reply:
Awesome, thanks a lot for helping out here! 😊